File: frontend/lib/api/roadmap.ts
import { API_BASE, getApiHeaders } from "./client";

export interface RoadmapMessage {
  id: string;
  session_id: string;
  role: "user" | "assistant" | "system";
  content: string;
  provider?: string;
  metadata_json?: any;
  created_at: string;
}

export interface RoadmapSessionListItem {
  id: string;
  title: string;
  target_role?: string;
  company?: string;
  created_at: string;
  updated_at: string;
  message_count: number;
}

export interface RoadmapSession {
  id: string;
  title: string;
  target_role?: string;
  company?: string;
  jd_text: string;
  jd_analysis?: any;
  created_at: string;
  updated_at: string;
  messages: RoadmapMessage[];
}

export async function createRoadmapSession(payload: {
  jd_text: string;
  title?: string;
}): Promise<RoadmapSession> {
  const res = await fetch(`${API_BASE}/api/roadmap/sessions`, {
    method: "POST",
    headers: getApiHeaders({ "Content-Type": "application/json" }),
    body: JSON.stringify(payload),
  });

  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail || "Failed to create roadmap session");
  }
  return res.json();
}

export async function getRoadmapSessions(): Promise<RoadmapSessionListItem[]> {
  const res = await fetch(`${API_BASE}/api/roadmap/sessions`, {
    headers: getApiHeaders(),
    cache: "no-store",
  });

  if (!res.ok) {
    return [];
  }
  return res.json();
}

export async function getRoadmapSession(sessionId: string): Promise<RoadmapSession> {
  const res = await fetch(`${API_BASE}/api/roadmap/sessions/${sessionId}`, {
    headers: getApiHeaders(),
    cache: "no-store",
  });

  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail || "Failed to load roadmap session");
  }
  return res.json();
}

export async function sendRoadmapMessage(
  sessionId: string,
  message: string
): Promise<RoadmapMessage> {
  const res = await fetch(`${API_BASE}/api/roadmap/sessions/${sessionId}/messages`, {
    method: "POST",
    headers: getApiHeaders({ "Content-Type": "application/json" }),
    body: JSON.stringify({ message }),
  });

  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail || "Failed to send roadmap message");
  }
  return res.json();
}

export async function deleteRoadmapSession(
  sessionId: string
): Promise<{ status: string; id: string }> {
  const res = await fetch(`${API_BASE}/api/roadmap/sessions/${sessionId}`, {
    method: "DELETE",
    headers: getApiHeaders(),
  });

  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail || "Failed to delete roadmap session");
  }
  return res.json();
}

File: backend/app/schemas/roadmap.py
from pydantic import BaseModel, Field
from typing import Optional, List, Any, Dict
from datetime import datetime


class RoadmapSessionCreate(BaseModel):
    jd_text: str = Field(..., min_length=10, description='Full or excerpted Job Description text')
    title: Optional[str] = Field(None, description='Optional session title')


class RoadmapMessageCreate(BaseModel):
    message: str = Field(..., min_length=1, description='Follow-up message or question')


class RoadmapMessageResponse(BaseModel):
    id: str
    session_id: str
    role: str
    content: str
    provider: Optional[str] = 'groq'
    metadata_json: Optional[Dict[str, Any]] = None
    created_at: datetime

    class Config:
        from_attributes = True


class RoadmapSessionListItem(BaseModel):
    id: str
    title: str
    target_role: Optional[str] = None
    company: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    message_count: int = 0

    class Config:
        from_attributes = True


class RoadmapSessionResponse(BaseModel):
    id: str
    title: str
    target_role: Optional[str] = None
    company: Optional[str] = None
    jd_text: str
    jd_analysis: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    messages: List[RoadmapMessageResponse] = []

    class Config:
        from_attributes = True
